File: clients/python/relais_client.py
__author__ = 'schinken'

import urllib2
import base64
import json

class restlib2(urllib2.Request):

    """ Workaround for using DELETE with urllib2
        Thanks to: http://abhinandh.com/post/2383952338/making-a-http-delete-request-with-urllib2
    """

    def __init__(self, url, method, data=None, headers={}, origin_req_host=None, unverifiable=False):
        self._method = method
        urllib2.Request.__init__(self, url, data, headers, origin_req_host, unverifiable)

    def get_method(self):

        if self._method:
            return self._method
        else:
            return urllib2.Request.get_method(self)


class RelaisClient(object):

    host = ''
    username=None
    password=None

    get = 'GET'
    set = 'POST'
    reset = 'DELETE'

    def __init__(self, host, port=80, protocol='https', username=None, password=None):
        self.host = "%s://%s:%s" % (protocol, host, port)
        self.username = username
        self.password = password

    def sendCommand(self, path, type ):

        try:
            req = restlib2( self.host+path, type )
            
            if self.username and self.password:
                base64string = base64.encodestring('%s:%s' % (self.username, self.password))[:-1]
                authheader = "Basic %s" % base64string
                req.add_header("Authorization", authheader)

            f = urllib2.urlopen( req )

        except ValueError:
            raise Exception('Unable to fulfill request')

        try:
            return json.load(f)
        except ValueError:
            raise Exception('Unable to parse result')

    def setPort(self, port, value=1 ):

        if not value:
            return self.resetPort( port )

        return self.sendCommand( "/relais/%d" % port, self.set )

    def setPorts(self, value=1 ):

        if not value:
            return self.resetPorts()

        return self.sendCommand("/relais", self.set )


    def resetPort(self, port):
        return self.sendCommand( '/relais/%d' % port, self.reset )

    def resetPorts(self):
        return self.sendCommand( '/relais', self.reset )


    def getPort(self, port):
        return self.sendCommand( '/relais/%d' % port, self.get )


    def getPorts(self):
        return self.sendCommand( '/relais', self.get )


if __name__ == '__main__':

    rc = RelaisClient(host='10.1.20.10', port=5000, protocol='http', username='john', password='doe2')
    rc.resetPort(3)

File: clients/php/relais_client.php
<?php

class RelaisClient {

    const SET   = 'POST';
    const GET   = 'GET';
    const RESET = 'DELETE';

    const ON    = 1;
    const OFF   = 0;

    private $Host = '';

    /**
     * @param $Host
     * @param int $Port
     */

    public function __construct( $Host, $Port = 80 ) {
        $this->Host = sprintf("http://%s:%d", $Host, $Port );
    }

    /**
     * Sends command to REST-API
     *
     * @param string $Path
     * @param string $Type
     * @return mixed
     * @throws Exception
     */

    private function sendCommand( $Path, $Type ) {

        
        $ch = curl_init( $this->Host . $Path );

        curl_setopt( $ch, CURLOPT_RETURNTRANSFER,       true );
        curl_setopt( $ch, CURLOPT_USERAGENT,            'PHP RelaisClient v1');
        curl_setopt( $ch, CURLOPT_CUSTOMREQUEST,        $Type );

        $result = curl_exec( $ch );
        if( $result ) {

            $decode = @json_decode( $result, true );
            if( $decode === null ) {
                throw new Exception('Unable to parse result');    
            }

            return $decode;

        } else {

            throw new Exception('Unable to fullfill request');
        }

    }

    /**
     * Set Port to On or Off
     *
     * @param int $Port
     * @param int $Value
     * @return bool
     */

    public function setPort( $Port, $Value = 1 ) {

        if( $Value == 0 ) {
            return $this->resetPort( $Port );    
        }
        
        return $this->sendCommand( sprintf('/ports/%d', $Port ), RelaisClient::SET );
    }

    /**
     * Sets all Ports to On or Off
     *
     * @param int $Value
     * @return bool
     */

    public function setPorts( $Value = 1 ) {

        if( $Value == 0 ) {
            return $this->resetPorts();    
        }
        
        return $this->sendCommand( '/ports', RelaisClient::SET );
    }

    /**
     * Resets Port (set to Off)
     *
     * @param int $Port
     * @return bool
     */

    public function resetPort( $Port ) {
        return $this->sendCommand( sprintf('/ports/%d', $Port ), RelaisClient::RESET );
    }

    /**
     * Reset ALL THE PORTS
     *
     * @return bool
     */

    public function resetPorts() {
        return $this->sendCommand( '/ports', RelaisClient::RESET );
    }

    /**
     * Returns value of the Port
     *
     * @param $Port
     * @return bool
     */

    public function getPort( $Port ) {
        return $this->sendCommand( sprintf('/ports/%d', $Port ), RelaisClient::GET );
    }

    /**
     * Returns state of all ports
     *
     * @return bool
     */

    public function getPorts() {
        return $this->sendCommand( '/ports', RelaisClient::GET );
        
    }
    
    
}

